[Refactor] 플레이스 리팩토링

hyobin-yang opened the pull request:
#130 

- CategoryCreateRequest, CategoryUpdateRequest를 삭제하고 CategoryTypeRequest로 통일

<br>

- MagazineCreateRequest, MagazineUpdateRequest를 삭제하고 MagazineRequest로 통일

<br>

- PlaceCreateRequest dto 내 모든 필드가 not null 제한 한 번에 걸 수 있도록 커스텀 애노테이션 구현(@AllFieldsNotNull)
   - AllFieldsNotNullValidator가 애노테이션의 검증기
   - 커스텀 애노테이션 작동은 테스트 코드로 정상 작동 확인 완료

<br>

- NPE 방지 위한 예외처리 코드 추가

hyobin-yang committed:
[refactor] 플레이스 로직 리팩토링

hyobin-yang committed:
[test] PlaceServiceImplTest 수정

yoonjiseok commented on src/main/java/com/pickyfy/pickyfy/service/PlaceServiceImpl.java:
```diff
@@ -77,32 +73,22 @@ public List<PlaceSearchResponse> getUserSavePlace(String email) {
-                            .naverLink(place.getNaverplaceLink())
-                            .iconUrl(savedMagazine.map(Magazine::getIconUrl).orElse(null))
-                            .build();
+                    return PlaceSearchResponse.from(new PlaceSearchResponseParams(
+                                    place,
+                                    placeImagesUrl,
+                                    savedCategory.map(Category::getName).orElse(null),
+                                    savedMagazine.map(Magazine::getTitle).orElse(null),
+                                    savedMagazine.map(Magazine::getIconUrl).orElse(null))
+                    );
```
따봉 드리겠습니다.

yoonjiseok commented on src/main/java/com/pickyfy/pickyfy/web/dto/response/PlaceSearchResponse.java:
```diff
@@ -1,26 +1,50 @@
+    private String naverLink;
+    private String iconUrl;
+
+    public static PlaceSearchResponse from(PlaceSearchResponseParams params){
+        Place place = params.place();
+        return PlaceSearchResponse.builder()
+                .placeId(place.getId())
+                .name(place.getName())
+                .shortDescription(place.getShortDescription())
+                .latitude(place.getLatitude())
+                .longitude(place.getLongitude())
+                .createdAt(place.getCreatedAt())
+                .updatedAt(place.getUpdatedAt())
+                .placeImageUrl(params.placeImageUrls())
+                .categoryName(params.categoryName())
+                .magazineTitle(params.magazineTitle())
+                .instagramLink(place.getInstagramLink())
+                .naverLink(place.getNaverplaceLink())
+                .iconUrl(params.iconUrl())
+                .placeImageId(place.getPlaceImages().stream().map(PlaceImage::getId).toList())
+                .build();
+    }
 }
```
좋은 방법 알아갑니다~

hyobin-yang committed:
[refactor] 메서드의 활용도를 높여보자

sese2204 commented on src/main/java/com/pickyfy/pickyfy/service/PlaceServiceImpl.java:
```diff
@@ -111,61 +89,34 @@ public PlaceSearchResponse getPlace(Long placeId) {
+                category.getName(),
+                magazine.getTitle(),
+                magazine.getIconUrl())
+        );
```
파라미터 많아서 싫다고 하셨는데 이런 방식이면 파라미터 똑같이 5개 아닌가요?
뭔가 괜히 new PlaceSearchResponseParams만 추가된 느낌인데 어떠신가요

hyobin-yang replied:
<img width="841" alt="image" src="https://github.com/user-attachments/assets/3c009794-f343-48ac-a89c-d65b941d9e62" />

넹 여기서 말했던 것과 같이 그 지점에 대해서도 고민했고 일단은 처음 생각한 대로 구현했습니다. 이유는 서비스 계층에서는 어쨌든 파라미터가 5개로 똑같지만 dto 내부에서는 PlaceSearchResponse 객체를 생성할 때 파라미터가 묶여서 하나로 들어갔으면 좋겠다는 생각이었습니다. 근데 아마 대부분 오빠처럼 불필요하다고 생각할 것 같어요 저도 어느정도 과하다는 거에 동의하고!
요 부분은 다시 바꾸겠습니당

sese2204 replied:
아뇨 다시 바꾸기보단 PlaceSearchResponseParams 조립을 따로 서비스에서 진행하고 from 메소드에는 조립된 객체만 넣는 식으로 하면 가독성이 훨씬 좋을 듯 해요

```
new PlaceSearchResponseParams(필요한 필드들..); 
return PlaceSearchResponse.from(params);
```

저도 서비스 전용 dto를 만드는 것을 선호해서 좋은 방법이라 생각합니다. 본인 의견 밀고 나가세요!

hyobin-yang replied:
아하 그런 의미였군요
근데 이미 바꿨어요 푸하하

hyobin-yang replied:
다시 만듦

sese2204 commented on src/main/java/com/pickyfy/pickyfy/service/PlaceServiceImpl.java:
```diff
@@ -242,39 +191,24 @@ public List<PlaceSearchResponse> getAllPlaces() {
-                    String magazineTitle = (placeMagazine != null) ?
-                            placeMagazine.getMagazine().getTitle() : "매거진 없음";
+                    String magazineTitle = (placeMagazine != null) ? placeMagazine.getMagazine().getTitle() : "매거진 없음";
 
```
null값 내려주는 것보다 String으로 "매거진 없음" 주는 것이 클라이언트에서 더 처리하기 편할까요?

그리고 매거진 정보 응답에 추가하는 부분에서 제가 dto에 MagazineInfo 추가해서 주변 장소 탐색에 사용하고 있는데,
해당 dto 사용하는 건 어때요

hyobin-yang replied:
오호 그러네요 내려줄 거면 Null값을 반환하는 게 맞는 것 같습니다.

근데 하나 의문인 게 현재 우리 서비스는 place 객체의 카테고리와 매거진 정보가 반드시 유효한 값(null X)이어야 하는 구조 아닌가 하는 겁니다. 맞다면 아예 예외를 던져버리는 게 맞는 게 아닌가 하는 생각이 드네요.

hyobin-yang replied:
오 MagazineInfo 있는지 지금 알았어요 완전 굿이다

hyobin-yang committed:
[refactor]
- CategoryCreateRequest, CategoryUpdateRequest를 삭제하고 CategoryTypeRequest로 통일
- PlaceCreateRequest dto 내 모든 필드가 not null이도록 커스텀 애노테이션 구현
- NPE 방지 위한 예외처리 코드 추가

hyobin-yang committed:
[feature] 전체 필드 null 검증 커스텀 애노테이션 구현

hyobin-yang committed:
[test] 커스텀 애노테이션 테스트 코드 추가

hyobin-yang committed:
[test] 커스텀 애노테이션 테스트 코드 추가

hyobin-yang committed:
[refactor] 매거진 중복 dto 삭제, MagazineRequest로 통일

hyobin-yang committed:
[refactor] PlaceSearchResponseParams 다시 생성

seungwon7934 commented on src/main/java/com/pickyfy/pickyfy/service/PlaceServiceImpl.java:
```diff
@@ -239,46 +170,28 @@ public void deletePlace(Long placeId) {
                 .collect(Collectors.toList());
     }
 
+    private PlaceSearchResponseParams createPlaceSearchResponseParams(Place place, Category category, Magazine magazine){
+        return new PlaceSearchResponseParams(
+                place,
+                placeImageRepository.findAllByPlaceId(place.getId()),
+                category.getName(),
+                new MagazineInfo(magazine.getTitle(), magazine.getIconUrl())
+        );
+    }
```
깔-끔 굿

seungwon7934 commented on src/test/java/com/pickyfy/pickyfy/common/util/AllFieldsNotNullValidatorTest.java:
```diff
@@ -0,0 +1,56 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AllFieldsNotNullValidatorTest {
```
요거는 Validator만 확인하는 클래스라 그냥 TestClass 쓴거라고 이해하면 되나요

hyobin-yang replied:
넹 그냥 간단한 테스트용 데이터 객체 하나 만들었어여

hyobin-yang committed:
[refactor] 최종 리팩토링
